### ISSUE TITLE
COPS - 503 Error

### ISSUE THREAD

**Issue by jlaidlaw86** [SynoCommunity/spksrc#5832]
### Is this a new Bug?

- [X] I checked that the bug hasn't been reported before

### Package Name

COPS

### Package Version

1.1.3-6

### Device Model

Synology DS418

### Device Architecture

AArch64 (ARMv8)

### Firmware Version

7.2-64570

### What happened?

 I Installed COPS and dependencies but i get  a 503 error when i try to launch the web app

### Reproduction steps

1. I Installed COPS and all dependencies
2. I launch the app
3. I get a 503 error
...

### Install Log

```log
I am accessing my Synology Device from a web browser and I appear to have very limited access to the file system. I am using File Station and /var/log/packages is not a valid pathway. I will gladly attach a log file if someone can tell me where to get it from
```

**Comment by hgy59** [comment]
@jlaidlaw86 thanks for reporting. Probably this issue is retaed to DSM 7.2 (synology is restricting permissions for thirdparty packages since DSM 7 and may be DSM 7.2 goes one step further.

It is very important that we get the installer and package log files located on the DSM unter
- `/var/log/packages/cops.log`
- `/var/packages/cops/var/cops.log`

You can access those file only on a shell and the most common way is to login via ssh (enable ssh service in control-panel under terminal & snmp).

A workaround without ssh access is to create a scheduled task (in control panel) and create a one time job that runs a script and within the script copy those files to a location that is accessible by file-station.

And there is a better way. Install the package `File Browser` from SynoCommunity. With this you can navigate the whole file system on DSM. You can view the log files (the download within the File View does not work, you have to download it in the folder view).

**Comment by hgy59** [comment]
> And there is a better way. Install the package `File Browser` from SynoCommunity. With this you can navigate the whole file system on DSM. You can view the log files (the download within the File View does not work, you have to download it in the folder view).

In the file view in `File Browser` there is no download button but a save button, that fails with `403 forbidden` since you are not allowed to edit and save those log files on the diskstation.

**Comment by jlaidlaw86** [comment]
[cops.log](https://github.com/SynoCommunity/spksrc/files/12264457/cops.log)

Thank you so much! i installed file browser and it worked like a charm. I am so glad synocommunity exists because there are so few official apps

**Comment by hgy59** [comment]
from installation log
```
2023/08/02 13:04:10	Begin service_preinst
2023/08/02 13:04:10	Metadata.db cannot be found. Please verify that the Calibre directory was entered correctly.
2023/08/02 13:04:10	End service_preinst
```

Do you have a calibre db file in the Calibre library directory?
If you have, I suppose rather an issue with file permissions.

**Comment by hgy59** [comment]
Probably this is an old issue as reported in #2520

**Comment by hgy59** [comment]
@jlaidlaw86 the same issue was resolved in #4620 (permissions)

**Comment by jlaidlaw86** [comment]
ok ill have a look at 2520 and 4620 

do i need to edit a config file to point it to my calibre folder or something?

**Comment by jlaidlaw86** [comment]
thanks for the speedy replies by the way . this is great

**Comment by hgy59** [comment]
> do i need to edit a config file to point it to my calibre folder or something?

If you installed with the default settings, the folder is `/volume1/calibre` and the system internal user `sc-cops` should be configured with read/write access.

To copy the calibre db to this folder, you can use the shared folder as windows/macos share or you can use the file station to access the folder.
We have a wiki page for the [permission management](https://github.com/SynoCommunity/spksrc/wiki/Permission-Management) and you will have to configure permissions for your regular (DSM) user to access this shared folder.

**Comment by jlaidlaw86** [comment]
ok ill do this and report back soon

**Comment by hgy59** [comment]
@jlaidlaw86 while testing cops in #5817 I realized that you must create a shared folder for the calibre data before you install the package.
At time of installation of cops the folder (you chose in the wizard) must contain all the calibre data including the `metadata.db` file.

**Comment by jlaidlaw86** [comment]
I have a "shared" folder setup in the sense that it is a folder that is discoverable on my home network.

Is that what you mean by "shared"?

Sent from Surface Duo
________________________________

**Comment by jlaidlaw86** [comment]
I had to uninstall COPS and now it longer shows as a package in package center even though i have the synocommunity repo. why cant i see it ?

**Comment by hgy59** [comment]
@jlaidlaw86 do you still have the beta packages enabled?

**Comment by jlaidlaw86** [comment]
i have "show beta packages" checked under package manager settings

**Comment by hgy59** [comment]
So I have not a clue what's going wrong here.
But you can manually download and install the package for your Model

For DSM 7.x download the [7.0 noarch](https://packages.synocommunity.com/cops/6/cops.v6.f40000%5Bnoarch%5D.spk) package

**Comment by hgy59** [comment]
Ok, I think I got it.
AFAICR on DSM >= 7 the beta packages are not shown in the package center, only Synology beta packages are shown.

@smaarn I think it is time to remove the `BETA=1` for the COPS package in #5817

**Comment by smaarn** [comment]
> Ok, I think I got it.
> 
> AFAICR on DSM >= 7 the beta packages are not shown in the package center, only Synology beta packages are shown.
> 
> 
> 
> @smaarn I think it is time to remove the `BETA=1` for the COPS package in #5817

There is also an issue with the PHP version which is no longer available if I'm not mistaken. Point taken for the beta flag (especially since the .htaccess file is now migrated).

**Comment by hgy59** [comment]
> There is also an issue with the PHP version which is no longer available if I'm not mistaken.

On `DSM 7.2-64570 Update 3` Synology currently provides PHP packages for versions 7.3, 7.4, 8.0, 8.1 and 8.2
On `DSM 7.1.1-42962 Update 6` (I have a model that does not support DSM 7.2) Synology currently provides PHP packages for versions 7.3, 7.4 and 8.0
On `DSM 6.2.4-25556 Update 7` Synology currently provides PHP packages for versions 7.3 and 7.4

So PHP 7.4 suits for all DSM >= 6.

**Comment by smaarn** [comment]
> > There is also an issue with the PHP version which is no longer available if I'm not mistaken.
> 
> 
> 
> On `DSM 7.2-64570 Update 3` Synology currently provides PHP packages for versions 7.3, 7.4, 8.0, 8.1 and 8.2
> 
> On `DSM 7.1.1-42962 Update 6` (I have a model that does not support DSM 7.2) Synology currently provides PHP packages for versions 7.3, 7.4 and 8.0
> 
> On `DSM 6.2.4-25556 Update 7` Synology currently provides PHP packages for versions 7.3 and 7.4
> 
> 
> 
> So PHP 7.4 suits for all DSM >= 6.
> 
> 

If I'm not mistaken the currently supported version is PHP 7.2. PR is also upgrading that part.

**Comment by jlaidlaw86** [comment]
[cops.log](https://github.com/SynoCommunity/spksrc/files/12316536/cops.log)

I created a calibre "shared" folder (/volume1/calibre)
Using Calibre, I migrated my library (including metadata.db) to this folder
I did a fresh install of COPS (with the wizard pointing to /volume1/calibre)

But i notice in the attached log it still cannot find the metadata.db file and i get 503 error

I do have multiple PHP packages installed...including 7.4 and 8.2. 

does this still appear to be a folder/permissions issue? or something else?

**Comment by hgy59** [comment]
without ssh access it is difficult to analyze.
_which version of cops?_

with synoacltool you can check the permissions ({owner} is a place holder for the (windows) user that copied the files to the share)

```
# synoacltool  -get  /volume1/calibre/metadata.db
ACL version: 1
Archive: is_inherit,is_support_ACL
Owner: [{owner}(user)]
---------------------
         [0] user:sc-cops:allow:rwxpdDaARWc--:---- (level:1)
         [1] group:administrators:allow:rwxpdDaARWc--:---- (level:1)
```

It is required that the `sc-cops` user has access to the share.
If it does not, you need to add this (system internal user) to have access to the shared folder.

**Comment by jlaidlaw86** [comment]
Not sure how to do this. I just went to control panel -> User&Group -> create 

am I doing this the right way? the wizard is asking me to set a password

**Comment by hgy59** [comment]
> Not sure how to do this. I just went to control panel -> User&Group -> create
> 
> am I doing this the right way? the wizard is asking me to set a password

Please see our wiki page [Permission Management](https://github.com/SynoCommunity/spksrc/wiki/Permission-Management) on how to do this.

**Comment by jlaidlaw86** [comment]
Ok. I followed the wiki and added sc-cops to calibre shared folder. But my issue remains. Sorry to keep being a pain

**Comment by smaarn** [comment]
The version you are using is the one fetched from Synocommunity official repository, right?
If so it's actually built against PHP 7.2. In order to make it work you would need to update the file /var/packages/cops/conf/resource and switch the php backend version to 8 (instead of 6). You could need to restart webstation to make it work.
Alternatively there is the manually built binaries of the in progress PR mentioned earlier. I am afk so I can't be more helpful unfortunately.

**Comment by jlaidlaw86** [comment]
I am not sure if this answers the question. but i am using the attached. is this built against 7.2?
[cops.v6.f40000[noarch].zip](https://github.com/SynoCommunity/spksrc/files/12317240/cops.v6.f40000.noarch.zip)

**Comment by smaarn** [comment]
> I am not sure if this answers the question. but i am using the attached. is this built against 7.2?
> 
> [cops.v6.f40000[noarch].zip](https://github.com/SynoCommunity/spksrc/files/12317240/cops.v6.f40000.noarch.zip)
> 
> 

I just extracted it and can confirm it does use PHP 7.2.

**Comment by smaarn** [comment]
> PHP 7.2 doesnt appear to be an available package anymore in the official repo. 
> 
> 
> 
> Is 7.2 an absolute requirement? or can I follow your suggestion above and edit the file in /var/packages/cops/conf/resource to point to a different version? If i do edit this file...its just a simple matter of opening it in a text editor, correct?
> 
> 
> 
> If i do in fact need 7.2.... where can i get it?

It should work just fine with 7.3 or 7.4
Not sure where you could get a 7.2 version.
Getting back to the resource file, it's indeed a plain text (json) file which you could safely edit (beware of the changes you make though)
Unfortunately I don't have access to a DSM right now so I can't really validate the behaviour.

**Comment by hgy59** [comment]
@jlaidlaw86 you can edit any file within an spk package except the `package.tgz` file (for a modified package.tgz file, the checksum in the INFO file must be updated to match the modified file).

I use 7-zip to unpack an spk into a folder.
in this folder edit the INFO file, i.e. change `install_dep_packages` to replace PHP7.2 by PHP7.4
after this, select all files and folders within in the folder and create a tar archive with 7-zip and finally rename the archive from *.tar to *.spk.

You can make another change in the package:
In the file `scripts/service-setup` you can rename the function `service_preinst` to `validate_preinst`.
with this modification, you will get an error and the installation is cancelled, when the `metadata.db` file is not found in the given folder.

**Comment by jlaidlaw86** [comment]
I am now using ssh as root with VI editor

I stopped web station cops services

I edited the resource file by:
1. changing PHP backend from 6 to 8
2. adding "zlib" to php extensions list

saved and closed. Restarted webstation 
But now the cops package fails to launch

should i try uninstalling, modifying the spk file per hgy59's first suggestion, and reinstalling cops?

**Comment by mesa57** [comment]
Still getting a 503 error on DSM7 (not DSM6)

```
synoacltool  -get  /volume1/SH3/calibre/metadata.db
ACL version: 1
Archive: is_inherit,is_support_ACL
Owner: [guest(user)]
---------------------
         [0] user:sc-cops:allow:rwxpdDaARWcCo:---- (level:1)
         [1] group:http:allow:rwxpdDaARWcCo:---- (level:1)
         [2] group:administrators:allow:rwxpdDaARWc--:---- (level:2)
         [3] user:guest:allow:rwxpdDaARWc--:---- (level:2)

```

**Comment by hgy59** [comment]
> Still getting a 503 error on DSM7 (not DSM6)
> 
> ```
> synoacltool  -get  /volume1/SH3/calibre/metadata.db
> ```

This is not a shared folder created in the control panel in DSM.
Shared folders are located directly under `/volume?/` like `/volume1/calibre/`

**Comment by mesa57** [comment]
Same issue
```
synoacltool  -get  /volume1/calibre/metadata.db
ACL version: 1
Archive: is_inherit,is_support_ACL
Owner: [guest(user)]
---------------------
         [0] group:administrators:allow:rwxpdDaARWc--:---- (level:1)
         [1] user:sc-cops:allow:rwxpdDaARWc--:---- (level:1)
         [2] user:guest:allow:rwxpdDaARWc--:---- (level:1)

```

**Comment by hgy59** [comment]
@jlaidlaw86, @mesa57  this is probably fixed with COPS 1.4.2-7.

**Comment by mreid-tt** [comment]
Hi @jlaidlaw86, it looks like this issue has been open for some time without recent updates. As indicated by @hgy59, a new version of COPS (v1.4.2-7) has been released which may have resolved this issue. Can you please test with the latest version and let us know if the issue persists? If we don't hear back from you in a few days, we'll assume the issue has been resolved and close this ticket. Thank you!
